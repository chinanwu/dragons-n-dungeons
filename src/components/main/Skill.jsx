import PropTypes from 'prop-types';
import React, { useCallback, useState } from 'react';

import Checkbox from '../general/Checkbox.jsx';

import './Skill.less';

export const Skill = ({ skill, modifier, proficiencyBonus }) => {
  const [isProficient, setIsProficient] = useState(false);

  const handleChange = useCallback(
    checked => {
      setIsProficient(checked);
    },
    [setIsProficient]
  );

  const calculateModifier = () =>
    isProficient ? modifier + proficiencyBonus : modifier;

  return (
    <div className="Skill">
      <Checkbox checked={isProficient} onChange={handleChange} />
      <div className="Skill__modifier">{calculateModifier()}</div>
      <div className="Skill__name">{skill}</div>
    </div>
  );
};

Skill.propTypes = {
  name: PropTypes.string,
  modifier: PropTypes.number,
  proficiencyBonus: PropTypes.number,
};

export default Skill;
