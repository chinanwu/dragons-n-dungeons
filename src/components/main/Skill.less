.Skill {
  display: flex;
}

.Skill input {
  width: auto;
}

.Skill__modifier {
  margin-left: 3px;
}

.Skill__name {
  margin-left: 6px;
}

@media all and (max-width: 1050px) {
  .Skill {
    margin-left: 6px;
  }
}
