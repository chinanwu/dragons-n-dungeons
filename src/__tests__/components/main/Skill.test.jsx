import { shallow } from 'enzyme';
import React, { useState } from 'react';

import { Skill } from '../../../components/main/Skill.jsx';

jest.unmock('../../../components/main/Skill.jsx');

describe('Skill component', () => {
  describe('rendering', () => {
    it('renders Skill', () => {
      const wrapper = shallow(<Skill skill={'test'} />);
      expect(wrapper.getElement()).toMatchSnapshot();
    });
  });

  describe('behaviour', () => {
    it('adds proficiency bonus when proficient', () => {
      useState.mockReturnValue([true]);
      const wrapper = shallow(<Skill skill={'test'} />);
      expect(wrapper.getElement()).toMatchSnapshot();
    });

    it('sets proficiency when checked', () => {
      const setIsProficient = jest.fn();
      useState.mockReturnValue([false, setIsProficient]);
      const event = true;
      const wrapper = shallow(<Skill skill={'test'} />);
      expect(wrapper.getElement()).toMatchSnapshot();
      wrapper.find('#testCheckbox').simulate('change', event);
      expect(setIsProficient.mock.calls).toEqual([[true]]);
    });
  });
});
