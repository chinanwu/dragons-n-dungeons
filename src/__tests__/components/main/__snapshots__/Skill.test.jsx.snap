// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Skill component behaviour adds proficiency bonus when proficient 1`] = `
<div
  className="Skill"
>
  <Checkbox
    ariaLabelledBy="skillName"
    checked={true}
    id="testCheckbox"
    onChange={[Function]}
  />
  <div
    className="Skill__modifier"
  >
    NaN
  </div>
  <div
    className="Skill__name"
    id="skillName"
  >
    test
  </div>
</div>
`;

exports[`Skill component behaviour sets proficiency when checked 1`] = `
<div
  className="Skill"
>
  <Checkbox
    ariaLabelledBy="skillName"
    checked={false}
    id="testCheckbox"
    onChange={[Function]}
  />
  <div
    className="Skill__modifier"
  />
  <div
    className="Skill__name"
    id="skillName"
  >
    test
  </div>
</div>
`;

exports[`Skill component rendering renders Skill 1`] = `
<div
  className="Skill"
>
  <Checkbox
    ariaLabelledBy="skillName"
    checked={false}
    id="testCheckbox"
    onChange={[Function]}
  />
  <div
    className="Skill__modifier"
  />
  <div
    className="Skill__name"
    id="skillName"
  >
    test
  </div>
</div>
`;
